test/threads_exec.jl

```diff
@@ -1312,4 +1313,231 @@ end
+            t = Threads.@spawn begin
+                cpu_time = Base.Experimental.task_cpu_time_ns(current_task())
+                wall_time = Base.Experimental.task_wall_time_ns(current_task())
+                for _ in 1:5
+                    x = time_ns()
+                    while time_ns() < x + 100
+                    end
+                    new_cpu_time = Base.Experimental.task_cpu_time_ns(current_task())
+                    new_wall_time = Base.Experimental.task_wall_time_ns(current_task())
+                    @test new_cpu_time > cpu_time
+                    @test new_wall_time > wall_time
+                    cpu_time = new_cpu_time
+                    wall_time = new_wall_time
+                end
+                last_cpu_time[] = cpu_time
+                last_wall_time[] = wall_time
+            end
+            wait(t)
+            final_cpu_time = Base.Experimental.task_cpu_time_ns(t)
+            final_wall_time = Base.Experimental.task_wall_time_ns(t)
+            @test last_cpu_time[] < final_cpu_time
+            @test last_wall_time[] < final_wall_time
+            # ensure many more tasks are run to make sure the counters are
+            # not being updated after a task is done e.g. only when a new task is found
+            @sync for _ in 1:Threads.nthreads()
+                Threads.@spawn rand()
+            end
+            @test final_cpu_time == Base.Experimental.task_cpu_time_ns(t)
+            @test final_wall_time == Base.Experimental.task_wall_time_ns(t)
+        finally
+            Base.Experimental.task_metrics(false)
+        end
+    end
+end
+
+@testset "task time counters: lots of spawns" begin
+    using Dates
+    try
+        Base.Experimental.task_metrics(true)
+        # create more tasks than we have threads.
+        # - all tasks must have: cpu time <= wall time
+        # - some tasks must have: cpu time < wall time
+        # - summing across all tasks we must have: total cpu time <= available cpu time
+        n_tasks = 2 * Threads.nthreads(:default)
+        cpu_times = Vector{UInt64}(undef, n_tasks)
+        wall_times = Vector{UInt64}(undef, n_tasks)
+        start_time = time_ns()
+        @sync begin
+            for i in 1:n_tasks
+                start_time_i = time_ns()
+                task_i = Threads.@spawn peakflops()
+                Threads.@spawn begin
+                    wait(task_i)
+                    end_time_i = time_ns()
+                    wall_time_delta_i = end_time_i - start_time_i
+                    cpu_times[$i] = cpu_time_i = Base.Experimental.task_cpu_time_ns(task_i)
+                    wall_times[$i] = wall_time_i = Base.Experimental.task_wall_time_ns(task_i)
+                    # task should have recorded some cpu-time and some wall-time
+                    @test cpu_time_i > 0
+                    @test wall_time_i > 0
+                    # task cpu-time cannot be greater than its wall-time
+                    @test wall_time_i >= cpu_time_i
```

Comment:
😞 
```
  Expression: wall_time_i >= cpu_time_i
   Evaluated: 13247163681 >= 30917448428008467
```
on CI https://buildkite.com/julialang/julia-master/builds/41766#0192e2c2-ae14-4f9f-93dd-9ea96e39642d
```
threads                                           (1) |        started at 2024-10-31T16:21:10.844 on pid 646
Test Summary:                               |  Pass  Total   Time
threads_exec.jl with JULIA_NUM_THREADS == 1 | 40471  40471  47.4s
Test Summary:                               |  Pass  Total   Time
threads_exec.jl with JULIA_NUM_THREADS == 2 | 49137  49137  51.2s
Test Summary:                               |  Pass  Total   Time
threads_exec.jl with JULIA_NUM_THREADS == 4 | 50145  50145  55.6s
Test Failed at /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/test/threads_exec.jl:1521
  Expression: wall_time_i >= cpu_time_i
   Evaluated: 13247163681 >= 30917448428008467
task time counters: lots of spawns: Error During Test at /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/test/threads_exec.jl:1495
  Got exception outside of a @test
  TaskFailedException
      nested task error: There was an error during testing
  Stacktrace:
    [1] sync_end(c::Channel{Any})
      @ Base ./task.jl:601
    [2] macro expansion
      @ ./task.jl:634 [inlined]
    [3] macro expansion
      @ /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/test/threads_exec.jl:1507 [inlined]
    [4] macro expansion
      @ /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/stdlib/v1.12/Test/src/Test.jl:1707 [inlined]
    [5] macro expansion
      @ /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/test/threads_exec.jl:1496 [inlined]
    [6] macro expansion
      @ /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/stdlib/v1.12/Test/src/Test.jl:1707 [inlined]
    [7] top-level scope
      @ /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/test/threads_exec.jl:33
    [8] include(mod::Module, _path::String)
      @ Base ./Base.jl:590
    [9] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:329
   [10] _start()
      @ Base ./client.jl:558
```

Reply:
And another 
```
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-26ec30e9b7/share/julia/test/threads_exec.jl:1514
  Expression: cpu_time_i > 0
   Evaluated: 0 > 0
task time counters: lots of spawns: Error During Test at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-26ec30e9b7/share/julia/test/threads_exec.jl:1491
  Got exception outside of a @test
  TaskFailedException
      nested task error: There was an error during testing
  Stacktrace:
    [1] sync_end(c::Channel{Any})
      @ Base ./task.jl:601
    [2] macro expansion
      @ ./task.jl:634 [inlined]
    [3] macro expansion
      @ /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-26ec30e9b7/share/julia/test/threads_exec.jl:1503 [inlined]
```

Reply:
```
  Expression: cpu_time_i > 0
   Evaluated: 0 > 0
```

Hoping this one at least was because we were setting the `t._state` to done before the final update to `t.cpu_time_ns`, so there was a race where `wait(t)` could return (because `wait(t)` is `while !istaskdone(t)`) before we've set `t.cpu_time_ns` to a non-zero value. Moved the update to be in finish_task before we set the `_state`